File: validate-binary-search-tree/validate-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


bool isValidBST(struct TreeNode* root){
    if (root == NULL) {
        return true;
    }

    int minVal, maxVal; bool valid;
    checkNode(root, &minVal, &maxVal, &valid);
    return valid;
}

void checkNode(struct TreeNode* root, int* minVal, int* maxVal, bool* valid) {
    // root must be not null
    *minVal = root->val;
    *maxVal = root->val;
    *valid = true;

    int leftMinVal, leftMaxVal; bool leftValid = true;
    int rightMinVal, rightMaxVal; bool rightValid = true;

    // max / less than
    // the max value of the left subtree less than the node's key
    if (root->left != NULL) {
        checkNode(root->left, &leftMinVal, &leftMaxVal, &leftValid);
        if (leftMaxVal < root->val && leftValid) {
            *minVal = leftMinVal;
            *valid = true;
        } else {
            *valid = false;
            return;
        }
    }

    // min / greater than
    // the min value of the right subtree greater than the node's key
    if (root->right != NULL) {
        checkNode(root->right, &rightMinVal,  &rightMaxVal, &rightValid);
        if (rightMinVal > root->val && rightValid) {
            *maxVal = rightMaxVal;
            *valid = true;
        } else {
            *valid = false;
            return;
        }
    }
}

File: validate-binary-search-tree/validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
    if root == nil {
        return true
    }
    _, _, valid := checkNode(root)
    return valid
}

func checkNode(root *TreeNode) (minVal int, maxVal int, valid bool) {
   
    minVal = root.Val
    maxVal = root.Val
    valid = true

    if root.Left != nil {
        leftMinVal, leftMaxVal, leftValid := checkNode(root.Left)
        if leftMaxVal < root.Val && leftValid {
            minVal = leftMinVal
            valid = true
        } else {
            valid = false
            return
        }
    }

    if root.Right != nil {
        rightMinVal, rightMaxVal, rightValid := checkNode(root.Right)
        if rightMinVal > root.Val && rightValid {
            maxVal = rightMaxVal
            valid = true
        } else {
            valid = false
            return
        }
    }
    
    return
}
